spacy/tokens/doc.pyx

```diff
@@ -1462,6 +1462,146 @@ cdef class Doc:
                 remove_label_if_necessary(attributes[i])
                 retokenizer.merge(span, attributes[i])
 
+    def from_json(self, doc_json):
+        """Convert a JSON document generated by Doc.to_json() to a Doc.
+
+        vocab (Vocab): A storage container for lexical types.
+        RETURNS (Doc): A doc instance corresponding to the specified JSON representation.
+        """
+
+        ### Token-level properties ###
+
+        words = []
+        # Map annotation type IDs to their string equivalents.
+        annot_types = {t: self.vocab.strings[t].lower() for t in (POS, HEAD, DEP, LEMMA, TAG, MORPH)}
```

Comment:
I think "types" is a bit confusing as a name here and would prefer `attr` or `attr_id` for the attr symbols like `POS` (we have the type `attr_id_t` for these types in `attrs.pxd`).

```suggestion
        token_attrs = {
            t: self.vocab.strings[t].lower()
            for t in (POS, HEAD, DEP, LEMMA, TAG, MORPH)
        }
```